frame/vamm/src/lib.rs

```diff
@@ -0,0 +1,332 @@
+	// ----------------------------------------------------------------------------------------------------
+
+	// Pallets use events to inform users when important changes are made.
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Emitted after a successful call to the [`create`](Pallet::create) function.
+		Created { vamm_id: VammIdOf<T>, state: VammStateOf<T> },
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Runtime  Errors
+	// ----------------------------------------------------------------------------------------------------
+
+	// Errors inform users that something went wrong.
+	#[pallet::error]
+	pub enum Error<T> {
+		/// This error is thrown when the caller tries to set `base_asset_reserves` to zero.
+		BaseAssetReserveIsZero,
+		/// This error is thrown when the caller tries to set `quote_asset_reserves` to zero.
+		QuoteAssetReserveIsZero,
+		/// This error is thrown when the caller tries to set `peg_multiplier` to zero.
+		PegMultiplierIsZero,
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                                Hooks
+	// ----------------------------------------------------------------------------------------------------
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                         Genesis Configuration
+	// ----------------------------------------------------------------------------------------------------
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		pub vamm_count: VammIdOf<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self { vamm_count: Default::default() }
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {
+			VammCounter::<T>::put(self.vamm_count);
+		}
+	}
+
+	// ----------------------------------------------------------------------------------------------------
+	//                                           Vamm Trait
+	// ----------------------------------------------------------------------------------------------------
+
+	impl<T: Config> Vamm for Pallet<T> {
+		type VammId = VammIdOf<T>;
+		type Balance = BalanceOf<T>;
+
+		/// Creates a new virtual market.
+		///
+		/// # Overview
+		/// In order for the caller to create new markets, it has to request it
+		/// to the VAMM, which is responsible to keep track of all active
+		/// markets. The VAMM creates the market, inserts it in storage,
+		/// deposits a [`Created`](Event::<T>::Created) event on the
+		/// blockchain and returns the new market ID to the caller.
+		///
```

Comment:
Some things to think about: 

Isn't there a difference between a market (like a perpetuals market in the Clearing House, e.g., BTC-PERP) and the market making mechanism used by it? E.g., a Limit Order Book where market makers quote their own prices, or Perp-v2's vAMM where they set range orders in a Uniswap-v3 model, or Cap's where they delegate the quoting of prices completely to the vAMM. What I'm getting at is that the vAMM is not the market itself, IMO, just a part of it (albeit a crucial one). It's as if the vAMM was the accounting procedure that its owner (e.g., the ClearingHouse) uses to keep track of supply and demand and quote prices accordingly, without actually storing and transferring tokens. In that case, it might be better to explicitly say that we're creating and managing virtual **automated** market **makers**.

Another point to consider is that the pallet is not **the** VAMM, but rather protocol for creating and managing several of them. In that respect, it is similar to the `pools` or `vaults` pallets. Perhaps we might even rename it to the `vamms` pallet.

Reply:
I just realized that the pallets I mentioned use singular names (`pool` and `vault`) 😅

Reply:
Just reverted the change from `vamms` to `vamm` :sweat_smile: